### PR TITLE
feat: implement function locators

### PR DESCRIPTION
**Pull request by jrandolf-google**

### CODE REVIEW

File: test/src/locator.spec.ts [hunk 1]
```diff
@@ -647,4 +647,34 @@ describe('Locator', function () {
...
+          return new Promise<boolean>(resolve => {
+            return setTimeout(() => {
+              return resolve(true);
+            }, 100);
+          });
+        })
+        .wait();
+      await expect(result).resolves.toEqual(true);
+    });
+    it('should work with actions', async () => {
```
Comment (OrKoN) [inline test/src/locator.spec.ts:690]: what happens if the function locator does not return an HTML element?

File: packages/puppeteer-core/src/api/Page.ts [hunk 2]
```diff
@@ -841,8 +843,26 @@ export class Page extends EventEmitter {
...
+  /**
+   * Creates a locator for the provided function. See {@link Locator} for
+   * details and supported actions.
+   *
+   * @remarks
+   * Locators API is experimental and we will not follow semver for breaking
+   * change in the Locators API.
+   */
+  locator<Ret>(func: () => Awaitable<Ret>): Locator<Ret>;
+
```
Comment (OrKoN) [inline packages/puppeteer-core/src/api/Page.ts]:
```suggestion
```

File: test/src/locator.spec.ts (hunk 1 above)
Comment (jrandolf) [reply test/src/locator.spec.ts:690]: TypeScript ensures that `click` cannot happen for a non-Element locator.

File: packages/puppeteer-core/src/api/Page.ts [hunk 3]
```diff
@@ -832,7 +838,7 @@ export class Page extends EventEmitter {
   }
 
   /**
-   * Creates a locator for the provided `selector`. See {@link Locator} for
+   * Creates a locator for the provided function. See {@link Locator} for
```
Comment (OrKoN) [inline packages/puppeteer-core/src/api/Page.ts]: should this one still say selector?

File: docs/guides/locators.md [hunk 4]
```diff
@@ -1,9 +1,9 @@
 # Locators
 
-Locators is a new experimental API that combines `waitForSelector` and element
-actions in a single unit. In combination with additional precondition checks
-this allows locators to retry failed actions automatically leading to less flaky
-automation scripts.
+Locators is an innovative, experimental API that combines the functionalities of
```
Comment (OrKoN) [inline docs/guides/locators.md]:
```suggestion
Locators is a new, experimental API that combines the functionalities of
```

sounds too pompous, IMO

File: docs/guides/locators.md [hunk 5]
```diff
@@ -1,9 +1,9 @@
...
+waiting and actions into a unified unit. With additional precondition checks, it
```
Comment (OrKoN) [inline docs/guides/locators.md]:
```suggestion
waiting and actions. With additional precondition checks, it
```

"unified unit" does not sound too good, unit is already unified by definition.

File: docs/guides/locators.md [hunk 6]
```diff
@@ -12,7 +12,30 @@ in the Locators API.
...
+
+- Waits for the element to become
+  [visible](https://pptr.dev/api/puppeteer.elementhandle.isvisible/) or hidden.
+
+### Waiting for a function
+
+```ts
+await page
+  .locator(() => {
+    return new Promise(resolve => setTimeout(resolve, 500));
```
Comment (OrKoN) [inline docs/guides/locators.md]: let's write an example that actually locates something? I cannot think of a good example, perhaps, using an Intersection observer or some other API to locate an element based on custom logic?

**Review by OrKoN — APPROVED** [review]

**Pull request merged**